Source/Utils/MsgLogger/MsgLogger.cpp

```diff
@@ -0,0 +1,564 @@
+        serialized_gather_rank_msgs.data(),
+        size_serialized_gather_rank_msgs, gather_rank);
+
+    return serialized_gather_rank_msgs;
+}
+
+std::vector<char>
+Logger::compute_package_for_gather_rank(
+    const std::vector<char>& serialized_gather_rank_msgs,
+    const int gather_rank_how_many_msgs,
+    const std::map<Msg, int>& my_msg_map,
+    const bool is_gather_rank)
+{
+    if(!is_gather_rank){
+        auto package = std::vector<char>{};
+
+        //generates a copy of the message map
+        auto msgs_to_send = std::map<Msg, int>{my_msg_map};
+
+        // For each message of the "gather rank" store how many times
+        // the message has been emitted by the current ranks.
+        const auto gather_rank_msgs =
+            Logger::deserialize_msgs(serialized_gather_rank_msgs);
+        std::vector<int> gather_rank_msg_counters(gather_rank_how_many_msgs);
+        int counter = 0;
+        for (const auto& msg : gather_rank_msgs){
+            const auto pp = msgs_to_send.find(msg);
+            if (pp != msgs_to_send.end()){
+                gather_rank_msg_counters[counter] += pp->second;
+                // Remove messages already seend by "gather rank" from
+                // the messages to send back
+                msgs_to_send.erase(msg);
+            }
+            counter++;
+        }
+        put_in_vec(gather_rank_msg_counters, package);
+
+        // Add the additional messages seen by the current rank to the package
+        put_in(static_cast<int>(msgs_to_send.size()), package);
+        for (const auto& el : msgs_to_send)
+            put_in_vec<char>(
+                MsgWithCounter{el.first, el.second}.serialize(), package);
+
+        return package;
+    }
+
+    return std::vector<char>{};
+}
+
+std::pair<std::vector<char>, std::vector<int>>
+Logger::gather_all_data(
+    const std::vector<char>& package_for_gather_rank,
+    const int gather_rank, const int my_rank)
+{
+    auto package_lengths = std::vector<int>{};
+    auto all_data = std::vector<char>{};
+    auto displacements = std::vector<int>{};
+
+    if(gather_rank != my_rank){
+        amrex::ParallelDescriptor::Gather(
+            static_cast<int>(package_for_gather_rank.size()), gather_rank);
+        amrex::ParallelDescriptor::Gatherv(
+            package_for_gather_rank.data(),
+            package_for_gather_rank.size(),
+            all_data.data(),
+            package_lengths,
+            displacements,
+            gather_rank);
+    }
+    else{
+        const int zero_size = 0;
+        package_lengths =
+            amrex::ParallelDescriptor::Gather(zero_size, gather_rank);
+
+        // Compute displacements
+        displacements.resize(package_lengths.size());
+        std::partial_sum(package_lengths.begin(), package_lengths.end(),
+            displacements.begin());
+        std::rotate(displacements.rbegin(),
+            displacements.rbegin()+1,
+            displacements.rend());
```

Comment:
I will add a comment.
Basically, given (n1, n2, n3, n4) we need (0, n1, n1+n2, n1+n2+n3) but partial_sum gives us (n1,n1+n2, n1+n2+n3, n1+n2+n3+n4). Rotating this last vector by one is just shifting: (n1+n2+n3+n4,n1, n1+n2, n1+n2+n3). Then we just need to replace the first element with zero: (0,n1, n1+n2, n1+n2+n3)

Reply:
That sounds like an exclusive scan, no? You can maybe use `std::exlusive_scan` or use the one defined in amrex (which returns the total sum that you need below)?

Reply:
`std::exclusive_scan` is unfortunately C++17 (but we should soon be able to use it, since at some point WarpX will switch to C++17). Using the one defined in `amrex` is a very good idea (certainly it will be more readable than `rotate` :-D)

Reply:
`amrex::Gpu::exclusive_scan` will do it on GPU for gpu builds.  So it is probably not what you want.

Reply:
Ok, thanks. In this case, using `rotate` with a comment should be fine.

Reply:
I've added a comment